src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,146 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private int arraySize;
+    private Object[] elementData;
+    private int size;
+
+    public ArrayList() {
+        arraySize = 10;
+        elementData = new Object[arraySize];
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size >= arraySize) {
+            Object[] obj = elementData;
+            elementData = new Object[(int) (elementData.length + (elementData.length * 0.5))];
+            for (int i = 0; i < obj.length; i++) {
+                elementData[i] = obj[i];
+            }
+            arraySize = elementData.length;
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
```

Comment:
The index check logic is repeated in multiple methods. Consider moving it to a separate private method to reduce redundancy.